### PR TITLE
Some more ABI tests.

### PR DESCRIPTION
**Pull request by chriseth**
This was extracted from https://github.com/ethereum/solidity/pull/2433

### CODE REVIEW

File: test/libsolidity/ABIEncoderTests.cpp [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+			function f(uint) returns (uint) {
+				function(uint) external returns (uint)[1] memory h;
+				assembly { sstore(0, sub(0, 1)) mstore(h, sub(0, 1)) }
+				E(h[0], g);
+			}
+		}
+	)";
+	compileAndRun(sourceCode);
+	callContractFunction("f(uint256)");
+	REQUIRE_LOG_DATA(encodeArgs(string(24, 0xff), string(24, 0xff)));
```
Comment (axic) [inline test/libsolidity/ABIEncoderTests.cpp]:
```
C:\projects\solidity\test\libsolidity\ABIEncoderTests.cpp(248): error C2220: warning treated as error - no 'object' file generated [C:\projects\solidity\build\test\soltest.vcxproj]
C:\projects\solidity\test\libsolidity\ABIEncoderTests.cpp(248): warning C4309: 'argument': truncation of constant value [C:\projects\solidity\build\test\soltest.vcxproj]
```

**Comment by chriseth** [comment]
Tried to fix the error, let's see.

File: test/libsolidity/ABIEncoderTests.cpp [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+			event E(function(uint) external returns (uint), function(uint) external returns (uint));
+			function(uint) external returns (uint) g;
+			function f(uint) returns (uint) {
+				g = this.f;
+				E(this.f, g);
+			}
+		}
+	)";
+	compileAndRun(sourceCode);
+}
```
Comment (axic) [inline test/libsolidity/ABIEncoderTests.cpp]: This one here is lacking actual expectations.

File: test/libsolidity/ABIEncoderTests.cpp [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+BOOST_AUTO_TEST_CASE(external_function_cleanup)
+{
+	char const* sourceCode = R"(
+		contract C {
+			event E(function(uint) external returns (uint), function(uint) external returns (uint));
+			function(uint) external returns (uint) g;
+			function f(uint) returns (uint) {
+				function(uint) external returns (uint)[1] memory h;
+				assembly { sstore(0, sub(0, 1)) mstore(h, sub(0, 1)) }
```
Comment (axic) [inline test/libsolidity/ABIEncoderTests.cpp]: Perhaps a comment that this relies on `g` being at storage slot 0.

File: test/libsolidity/ABIEncoderTests.cpp [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+				int16 d;
+				assembly { a := sub(0, 1) c := 0x0101ff d := 0xff01 }
+				E(10, x, a, uint8(b), c, int8(d));
+			}
+		}
+	)";
+	compileAndRun(sourceCode);
+	callContractFunction("f()");
+	REQUIRE_LOG_DATA(encodeArgs(
+		string(3, 0) + string("\x0a"), string("\xf1\xf2"),
```
Comment (axic) [inline test/libsolidity/ABIEncoderTests.cpp:124]: Why not `string("\x00\x00\x00\x0a")` for the first one?

**Comment by axic** [comment]
Validated the encoding by hand, looks good.

File: test/libsolidity/ABIEncoderTests.cpp (hunk 4 above)
Comment (chriseth) [reply test/libsolidity/ABIEncoderTests.cpp:124]: because `"\x00\x00\x00\x0a"` is a c-string which is auto-truncated at the first zero byte.

**Comment by chriseth** [comment]
Updated

**Review by axic — APPROVED** [review]

**Pull request merged**